edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+        self.name = name
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+    def get_code_url(self) -> str:
+        raise NotImplementedError
+
+    async def exchange_code(self, code: str) -> str:
+        raise NotImplementedError
+
+    async def fetch_user_info(self, token: str) -> dict:
+        raise NotImplementedError
+
+
+# OAuth 2 Client HTTP Endpoints
+
+
+def redirect_to_auth_provider(
+    response,
+    redirect_uri: str,
+    iss: str,
+    provider_name: str,
+    db_config: SettingsMap,
+):
+    client_id, client_secret = _get_client_credentials(provider_name, db_config)
+    provider = _get_provider(
+        name=provider_name,
+        client_id=client_id,
+        client_secret=client_secret,
+    )
+
+    signing_key = _get_auth_signing_key(db_config)
+    state_token = _make_signed_token(
+        iss=iss, provider=provider_name, key=signing_key
+    )
+
+    auth_url = provider.get_code_url(
+        state=state_token.serialize(), redirect_uri=redirect_uri
+    )
+
+    response.status = http.HTTPStatus.FOUND
+    response.custom_headers["Location"] = auth_url
+    response.close_connection = True
+
+
+async def handle_auth_callback(request, response):
+    return
+
+
+async def handle_request(request, response, db, args):
+    try:
+        # Set up routing to the appropriate handler
+        if args[0] == "authorize":
+            provider_name: str | None = _get_search_param(
+                request.url.query.decode("ascii"), "provider"
+            )
+            if provider_name is None:
+                raise errors.InternalServerError("No provider specified")
+
+            redirect_uri = f"{_get_extension_path(request.url)}/callback"
+
+            return redirect_to_auth_provider(
+                response=response,
+                redirect_uri=redirect_uri,
+                iss=request.host.decode(),
+                provider_name=provider_name,
+                db_config=db.db_config,
+            )
+        elif args[0] == "callback":
+            await handle_auth_callback(request, response)
+        else:
+            raise errors.BackendError
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+
+        # XXX Fix this when LSP "location" objects are implemented
+        ex_type = errors.InternalServerError
+
+        _response_error(
+            response, http.HTTPStatus.INTERNAL_SERVER_ERROR, str(ex), ex_type
+        )
+
+
+def _response_error(response, status, message, ex_type):
+    err_dct = {
+        "message": message,
+        "type": str(ex_type.__name__),
+        "code": ex_type.get_code(),
+    }
+
+    response.body = json.dumps({"error": err_dct}).encode()
+    response.status = status
+    response.close_connection = True
+
+
+# OAuth 2 Provider for GitHub
+class GitHubProvider(BaseProvider):
+    def __init__(self, client_id, client_secret):
+        super().__init__("github", client_id, client_secret)
```

Comment:
```suggestion
    def __init__(self, *args, **kwargs):
        super().__init__("github", *args, **kwargs)
```

Reply:
Can you suggest good reading material on this? I _think_ I understand how this works by unpacking the `args` list and the `kwargs` dict, but just to fill in the holes in my understanding.

Reply:
I'd glance though the docs -- https://docs.python.org/3/tutorial/controlflow.html#more-on-defining-functions -- I just tried to google for a good blog post on this and they're all garbage. Python argument packing is pretty powerful (more so than in most other languages) so I suggest reading the docs fully just so that you know what's possible.